federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -858,4 +868,53 @@ public LDAPObject loadLDAPUserByUuid(RealmModel realm, String uuid) {
         }
     }
 
+    private Predicate<LDAPObject> filterLocalUsers(RealmModel realm) {
+        return ldapObject -> UserStoragePrivateUtil.userLocalStorage(session).getUserByUsername(realm, LDAPUtils.getUsername(ldapObject, LDAPStorageProvider.this.ldapIdentityStore.getConfig())) == null;
+    }
+
+    private Stream<LDAPObject> paginatedSearchLDAP(LDAPQuery ldapQuery, Integer firstResult, Integer maxResults) {
```

Comment:
Question: There is already some pagination support provided by `LDAPQuery.getResultsList()` - as that one should delegate to `LDAPIdentityStore.fetchQueryResults` and that one eventually delegates to `LDAPOperationManager.searchPaginated` . But I suppose you are aware of that and there is a reason why you introduced this private method here as the built-in pagination is missing something?

Reply:
This method leverages the existing pagination support in `LDAPQuery.getResultsList()`. It still delegates it to `LDAPIdentityStore.fetchQueryResults` and then to `LDAPOperationManager.searchPaginated`. It only sets necessary fields to the ldapQuery.

It was introduced to ensure we would be able to obtain required number of users from LDAP considering a fact that some of the returned ldap users could be filtered out (as they might be already imported in local storage).

This method will load additional ldapObjects "on demand".

Michal, as it was your idea, would you like to add something here?
@mhajas ^

I'll add a java-doc to this method.